server/src/main/java/org/opensearch/identity/noop/NoopPluginSubject.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.identity.noop;
+
+import org.opensearch.common.annotation.InternalApi;
+import org.opensearch.common.util.concurrent.ThreadContext;
+import org.opensearch.identity.NamedPrincipal;
+import org.opensearch.identity.PluginSubject;
+import org.opensearch.threadpool.ThreadPool;
+
+import java.security.Principal;
+import java.util.concurrent.Callable;
+
+/**
+ * Implementation of subject that is always authenticated
+ * <p>
+ * This class and related classes in this package will not return nulls or fail permissions checks
+ *
+ * This class is used by the NoopIdentityPlugin to initialize IdentityAwarePlugins
+ *
+ * @opensearch.internal
+ */
+@InternalApi
+public class NoopPluginSubject implements PluginSubject {
+    private final ThreadPool threadPool;
+
+    NoopPluginSubject(ThreadPool threadPool) {
+        super();
+        this.threadPool = threadPool;
+    }
+
+    @Override
+    public Principal getPrincipal() {
+        return NamedPrincipal.UNAUTHENTICATED;
+    }
+
+    @Override
+    public <T> T runAs(Callable<T> callable) throws Exception {
+        try (ThreadContext.StoredContext ctx = threadPool.getThreadContext().stashContext()) {
```

Comment:
wdyt about adding a line in here that populates a ThreadContext header with the canonical class name of the plugin like in previous iteration?

I'm concerned with the change we made behind experimental, because my plan was to have plugins start integrating with this change to verify that stronger system index protection does not break their plugin. Without a unique identifier that security can reference we would have to wait for IdentityPlugin to be out of experimental and then verify that stronger system index protection does not break plugins all at once instead of letting them be able to verify as they adopt `pluginSubject.runAs()` as a replacement to `stashContext`. If it helps, I will open up a Security PR on Monday to demonstrate how security will use this header to enforce that only transport actions that interact with the plugin's system index are allowed and other actions denied.

It would still effectively be a Noop, the only difference being that a single header would be set. I was thinking of submitting a follow-up PR to this one to introduce a `SystemSubject` that could be used to replace blocks like [this](https://github.com/opensearch-project/OpenSearch/blob/9661e8db62e8897cb53b0bc1860ead43b28dfa21/server/src/main/java/org/opensearch/index/seqno/GlobalCheckpointSyncAction.java#L100-L102) where the implementation of `runAs` could be:

```
@Override
public <T> T runAs(Callable<T> callable) throws Exception {
    try (ThreadContext.StoredContext ctx = threadPool.getThreadContext().stashContext()) {
         ThreadContextAccess.doPrivilegedVoid(threadContext::markAsSystemContext);
         callable.call();
    }
}
```

With SystemSubject the first line in the stashContext block would mark the context, so NoopPluginSubject would be similar where the first line within the stashContext block populates a special header. We could also consider renaming NoopIdentityPlugin to DefaultIdentityPlugin. Its still a Noop, but a slight change in name to suggest it provides a default implementation.

Reply:
> wdyt about adding a line in here that populates a ThreadContext header with the canonical class name of the plugin like in previous iteration?

-1, we do have a full fledged `Subject` here. The next step would be to make it part of the thread context in a pluggable way (as part of IdentityPlugin APIs). 

> I'm concerned with the change we made behind experimental, because my plan was to have plugins start integrating with this change to verify that stronger system index protection does not break their plugin. 

The experimental part gives us more freedom to make breaking changes: the APIs is not stable yet and will evolve. 

> Without a unique identifier that security can reference we would have to wait for IdentityPlugin to be out of experimental ...

Absolutely not - the experimental part applies to guarantees to keep changes non-breaking. Any experimental API could be used right away with such guarantees in mind.

> Edit: I'm thinking about adding getPluginSubject to IdentityService which returns the principal of the pluginSubject if the subject is "active" (within a runAs block), otherwise returning null or empty optional.

Yes, we need such mechanism, but it should be managed by `IdentityPlugin` APIs (we may use `IdentityService` internally if it makes sense but this service is not exposed and could not be used by plugin implementors).

Reply:
> Absolutely not - the experimental part applies to guarantees to keep changes non-breaking. Any experimental API could be used right away with such guarantees in mind.

It's not breaking though, `ThreadContext.stashContext()` is still available and not guarded by a JSM permission. Plugins can rely on this and opt into using `pluginSubject.runAs(() -> { ... })` after performing testing with this change.

With the changes in this PR as is, plugins will need to modify any testing they do with security to set `opensearch.experimental.feature.identity.enabled: true` to integrate with the change and ensure that functionality in the plugin works fine with [Stronger System Index Protection](https://github.com/opensearch-project/security/issues/4439)

Reply:
Ok, that makes sense as long as you think its appropriate that plugins that do CI checks w/ security set `opensearch.experimental.feature.identity.enabled` to `true` in `opensearch.yml`.

Reply:
> It's not breaking though, ThreadContext.stashContext() is still available and not guarded by a JSM permission. Plugins can rely on this and opt into using pluginSubject.runAs(() -> { ... }) after performing testing with this change.

I think there are 2 parts to this: 
 - the current `Subject` has to be available on per context basis (to any plugin / extension)
 - the plugin / extension may change the `Subject` on per context basis (using `runAs`)

> With the changes in this PR as is, plugins will need to modify any testing they do with security to set opensearch.experimental.feature.identity.enabled: true to integrate with the change and ensure that functionality in the plugin works fine with https://github.com/opensearch-project/security/issues/4439

Oh I see, I missed this part, we could probably promote the identity plugin out of experimental feature but keep all APIs experimental.